src/math/fft/fft_32_hifi3.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <andrula.song@intel.com>
+
+#include <sof/audio/format.h>
+#include <sof/common.h>
+#include <rtos/alloc.h>
+#include <sof/math/fft.h>
+
+#ifdef FFT_HIFI3
+#include <sof/audio/coefficients/fft/twiddle_32.h>
+#include <xtensa/tie/xt_hifi3.h>
+
+void fft_execute_32(struct fft_plan *plan, bool ifft)
+{
+	struct icomplex32 tmp1;
+	ae_int32x2 *inx;
+	ae_int32x2 *out;
+	ae_int32x2 *outtop;
+	ae_int32x2 *outbottom;
+	ae_int32x2 *outx;
+	ae_int32x2 sample;
+	ae_int32x2 sample1;
+	ae_int32x2 sample2;
+	ae_int64 res, res1;
+	int depth, top, bottom, index;
+	int i, j, k, m, n;
+	ae_int32 *in;
+	ae_valign inu = AE_ZALIGN64();
+	ae_valign outu = AE_ZALIGN64();
+	int size = plan->size;
+	int len = plan->len;
+
+	if (!plan || !plan->bit_reverse_idx)
+		return;
+
+	if (!plan->inb32 || !plan->outb32)
+		return;
+
+	inx = (ae_int32x2 *)plan->inb32 + 1;
+	outx = (ae_int32x2 *)plan->outb32;
+
+	/* convert to complex conjugate for ifft */
+	if (ifft) {
+		in = (ae_int32 *)&plan->inb32->imag;
+		for (i = 0; i < size; i++) {
+			AE_L32_IP(sample, in, 0);
+			sample = AE_NEG32S(sample);
+			AE_S32_L_IP(sample, in, sizeof(struct icomplex32));
+		}
+	}
+
+	/* step 1: re-arrange input in bit reverse order, and shrink the level to avoid overflow */
+	inu = AE_LA64_PP(inx);
+	for (i = 1; i < size; ++i) {
+		AE_LA32X2_IP(sample, inu, inx);
+		sample = AE_SRAA32S(sample, len);
+		out = (ae_int32x2 *)&outx[plan->bit_reverse_idx[i]];
```

Comment:
Is this type cast necessary? outx seems to be int32x2 type already.

Reply:
yeah, thanks